### PR TITLE
new process and PIDManager

### LINKED ISSUE #5: Make PID Manager Class

### LINKED ISSUE #6: Make Process Class

### PR DESCRIPTION
**Pull request by sagerbala**
PIDmanager and process

### CODE REVIEW

File: Scheduler/src/PIDManager.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import java.util.BitSet;
+
+public class PIDManager {
+    private final int minPid;
+    private final int maxPid;
```
Comment (Siri2K) [inline Scheduler/src/PIDManager.java]:
```suggestion
    private static final int minPid = 300;
    private static final int maxPid = 5000;
```

File: Scheduler/src/PIDManager.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import java.util.BitSet;
+
+public class PIDManager {
+    private final int minPid;
+    private final int maxPid;
+    private BitSet pidBitmap;
+
+    public PIDManager(int minPid, int maxPid) {
```
Comment (Siri2K) [inline Scheduler/src/PIDManager.java:8]:
```suggestion
    public PIDManager() {

```

File: Scheduler/src/PIDManager.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+public class PIDManager {
+    private final int minPid;
+    private final int maxPid;
+    private BitSet pidBitmap;
+
+    public PIDManager(int minPid, int maxPid) {
+        this.minPid = minPid;
+        this.maxPid = maxPid;
+        this.pidBitmap = new BitSet(maxPid - minPid + 1);
```
Comment (Siri2K) [inline Scheduler/src/PIDManager.java]:
```suggestion
        this.pidBitmap = new BitSet(this.maxPid - this.minPid + 1);

```

File: Scheduler/src/PIDManager.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        initializePidBitmap();
```
Comment (Siri2K) [inline Scheduler/src/PIDManager.java]:
```suggestion
        this.allocateMap();

```

File: Scheduler/src/PIDManager.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    // Creates and initializes a data structure for representing pids
+    public int allocateMap() {
+        initializePidBitmap();
+        return 1;
+    }
+
+    // Allocates and returns a pid; returns -1 if unable to allocate a pid (all pids are in use)
+    public int allocatePid() {
+        int pid = pidBitmap.nextClearBit(0);
+        if (pid <= maxPid) {
```
Comment (Siri2K) [inline Scheduler/src/PIDManager.java:24]:
```suggestion
        if (pid >= this.minPid && pid<= this.maxPid) {

```

File: Scheduler/src/PIDManager.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+            pidBitmap.set(pid);
+            return pid + minPid;
```
Comment (Siri2K) [inline Scheduler/src/PIDManager.java:26]:
```suggestion
            return pid;

```

File: Scheduler/src/PIDManager.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        } else {
+            return -1; // No available PID
+        }
+    }
+
+    // Releases a pid
+    public void releasePid(int pid) {
+        if (isValidPid(pid)) {
+            pidBitmap.clear(pid - minPid);
```
Comment (Siri2K) [inline Scheduler/src/PIDManager.java:37]:
```suggestion
            pidBitmap.clear(pid);

```

File: Scheduler/src/PIDManager.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    // Initialize the PID bitmap (all PIDs are initially available)
+    private void initializePidBitmap() {
+        pidBitmap.clear();
+    }
+
+    // Checks if a PID is within the valid range
+    private boolean isValidPid(int pid) {
+        return pid >= minPid && pid <= maxPid;
```
Comment (Siri2K) [inline Scheduler/src/PIDManager.java]:
```suggestion
        return pid >= this.minPid && pid <= this.maxPid;

```

File: Scheduler/src/Process.java [hunk 9]
```diff
@@ -1,22 +1,26 @@
-import java.util.ArrayList;
-import java.util.List;
+import java.util.*;
```
Comment (Siri2K) [inline Scheduler/src/Process.java:1]: *imports all libraries (including extra ones with similar functions) which could impact the performance of the program. Import the specific ones being used.

**Comment by Siri2K** [comment]
Please add Setters and Getters for all class (even if we don't use them)

File: Scheduler/src/PIDManager.java (hunk 7 above)
Comment (sagerbala) [reply Scheduler/src/PIDManager.java:37]: needed because pid does not represent index only the offset

**Review by Siri2K — APPROVED** [review]

**Pull request merged**